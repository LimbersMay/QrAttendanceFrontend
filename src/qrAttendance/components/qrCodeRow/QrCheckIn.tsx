import {Box, Dialog, DialogContent, DialogTitle, IconButton} from "@mui/material";
import {Close} from "@mui/icons-material";
import QRCode from 'react-qr-code';

export const QrCheckIn = ({ url, title, isQrShowing, handleToggleShowQr }: { url: string, title: string, isQrShowing: boolean, handleToggleShowQr: any}) => {
    return (
        <Dialog open={isQrShowing} maxWidth="sm" fullWidth>
            <Box
                display="flex"
                justifyContent="center"
            >
                <DialogTitle>{ title } QR Code</DialogTitle>
            </Box>
            <Box position="absolute" top={0} right={0}>
                <IconButton onClick={handleToggleShowQr}>
                    <Close />
                </IconButton>
            </Box>
            <DialogContent>
                <Box
                    sx={{height: 'auto', margin: "0 auto", maxWidth: "64", width: "100%"}}
                    display="flex"
                    justifyContent="center"
                >
                    <QRCode
                        size={100}
                        style={{ height: "auto", maxWidth: "50%", width: "80%" }}
                        value={url}
                        viewBox={`0 0 256 256`}
                    />
                </Box>
            </DialogContent>
        </Dialog>
    );
}
