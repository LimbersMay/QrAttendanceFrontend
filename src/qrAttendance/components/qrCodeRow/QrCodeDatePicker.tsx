import * as React from 'react';
import dayjs, { Dayjs } from 'dayjs';
import Stack from '@mui/material/Stack';
import TextField from '@mui/material/TextField';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
import { AdapterDayjs } from '@mui/x-date-pickers/AdapterDayjs';
import { DesktopDatePicker } from '@mui/x-date-pickers/DesktopDatePicker';

export const QrCodeDatePicker = ({ date, onChangeDate }: { date: string | undefined, onChangeDate: any }) => {
    const [value, setValue] = React.useState<Dayjs | null>(
        dayjs(date),
    );

    const handleChange = (newValue: Dayjs | null) => {
        console.log('date: ', value?.toISOString())
        setValue(newValue);
        onChangeDate(newValue?.toISOString());
    };

    return (
        <LocalizationProvider dateAdapter={AdapterDayjs}>
            <Stack spacing={1}>
                <DesktopDatePicker
                    label="Date"
                    inputFormat="DD/MM/YYYY"
                    value={value}
                    onChange={handleChange}
                    renderInput={(params) => <TextField {...params} />}
                />
            </Stack>
        </LocalizationProvider>
    );
}