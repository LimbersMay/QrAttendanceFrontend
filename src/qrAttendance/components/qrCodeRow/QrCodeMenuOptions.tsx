import React from 'react';
import {Divider, IconButton, Menu, MenuItem, Typography} from "@mui/material";
import {Delete, Download, Edit, MoreVert, Visibility} from "@mui/icons-material";
import {useQrCodeStore} from "../../../hooks/useQrCodeStore";
import {QrCode} from "../../interfaces";
import {useUiStore} from "../../../hooks/useUiStore";

const ITEM_HEIGHT = 48;

export const QrCodeMenuOptions = ({handleDownload, qrCode }: { handleDownload: any, qrCode: QrCode }) => {

    const { showQrCode, toggleQrCodeModal } = useUiStore();
    const { setActiveQrCode, startDeleteQrCode } = useQrCodeStore();

    const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
    const open = Boolean(anchorEl);

    const handleClick = (event: React.MouseEvent<HTMLElement>) => {
        setAnchorEl(event.currentTarget);
        setActiveQrCode(qrCode);
    };

    const handleClose = () => {
        setAnchorEl(null);
        setActiveQrCode(null);
    };

    return (
        <>
            <IconButton
                aria-label="more"
                aria-controls="long-menu"
                aria-haspopup="true"
                onClick={handleClick}
            >
                <MoreVert />
            </IconButton>
            <Menu
                id="long-menu"
                anchorEl={anchorEl}
                keepMounted
                open={open}
                onClose={handleClose}
                PaperProps={{
                    style: {
                        maxHeight: ITEM_HEIGHT * 4.5,
                        width: '20ch',
                    },
                }}
            >
                <MenuItem onClick={showQrCode}>
                    <Visibility />
                    <Typography sx={{ml: '7px'}}>Show</Typography>
                </MenuItem>

                <MenuItem onClick={toggleQrCodeModal}>
                    <Edit />
                    <Typography sx={{ml: '7px'}}>Edit</Typography>
                </MenuItem>

                <MenuItem onClick={handleDownload}>
                    <Download />
                    <Typography sx={{ml: '7px'}}>Download</Typography>
                </MenuItem>

                <Divider />

                <MenuItem onClick={startDeleteQrCode}>
                    <Delete />
                    <Typography sx={{ml: '7px'}}>Delete</Typography>
                </MenuItem>

            </Menu>
        </>
    );
}
