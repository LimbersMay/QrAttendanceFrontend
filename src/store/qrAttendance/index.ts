export * from "./group";
export * from "./qrCode";
export * from "./registry";
