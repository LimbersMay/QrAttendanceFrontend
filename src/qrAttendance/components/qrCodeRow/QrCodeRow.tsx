import {memo} from "react";

import dayjs from "dayjs";
import {IconButton, TableCell, TableRow} from "@mui/material";
import KeyboardArrowUpIcon from "@mui/icons-material/KeyboardArrowUp";

import KeyboardArrowDownIcon from "@mui/icons-material/KeyboardArrowDown";
import {QrCode, Registry} from "../../interfaces";
import {QrCodeMenuOptions} from "./QrCodeMenuOptions";
import {QrCheckIn} from "./QrCheckIn";
import {generateExcelFromRegistries} from "../../helpers/generateExcelFromRegistries";

export const QrCodeRow = memo(({
                                   qrCodeRow,
                                   handleOpenSubTable,
                                   open,
                                   registries
                               }: { qrCodeRow: QrCode, handleOpenSubTable: Function, open: boolean, registries: Registry[] }) => {


    const { name, date, enabled } = qrCodeRow;

    const handleDownload = () => {
        generateExcelFromRegistries(qrCodeRow,registries);
    }

    return (
        /* Rows of the table with QrCode components */
        <TableRow sx={{'& > *': {borderBottom: 'unset'}}}>
            <TableCell>
                <IconButton
                    aria-label="expand row"
                    size="small"
                    onClick={() => handleOpenSubTable(!open)}
                >
                    {open ? <KeyboardArrowUpIcon/> : <KeyboardArrowDownIcon/>}
                </IconButton>
            </TableCell>
            <TableCell component="th" scope="row">
                {name}
            </TableCell>
            <TableCell align="center">
                {registries.length}
            </TableCell>
            <TableCell align="center" sx={{display: {xs: 'none', sm: 'table-cell'}}}>
                {dayjs(date).format('DD/MM/YYYY')}
            </TableCell>
            <TableCell align="center">
                {enabled ? 'YES' : 'NO'}
            </TableCell>
            <TableCell align="center">
                <QrCodeMenuOptions
                    handleDownload={handleDownload}
                    qrCode={qrCodeRow}
                />
            </TableCell>

            <QrCheckIn/>
        </TableRow>
    )
});