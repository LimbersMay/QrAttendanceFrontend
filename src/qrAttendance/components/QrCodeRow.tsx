import {useState} from "react";

import {IconButton, MenuItem, Select, SelectChangeEvent, TableCell, TableRow} from "@mui/material";
import KeyboardArrowUpIcon from "@mui/icons-material/KeyboardArrowUp";
import KeyboardArrowDownIcon from "@mui/icons-material/KeyboardArrowDown";
import {Save} from "@mui/icons-material";

import {useForm} from "../../hooks/useForm";
import {ConditionalTextField} from "./ConditionalTextField";
import {QrCode} from "../interfaces";
import {useAppDispatch} from "../../store";
import { startDeleteQrCodeWithDependencies, startUpdateQrCode} from "../../store/qrAttendance";
import {QrCodeMenuOptions} from "./QrCodeMenuOptions";

export const QrCodeRow = ({
           qrCodeRow,
           handleOpenSubTable,
           open,
    registriesLength
}:{ qrCodeRow: QrCode, handleOpenSubTable: Function, open: boolean, registriesLength: number }) => {

    const dispatch = useAppDispatch();

    const [isEditing, setIsEditing] = useState<boolean>(false);
    const [isEnable, setIsEnable] = useState<boolean>(qrCodeRow.enabled);

    const {formState, onInputChange} = useForm(qrCodeRow);
    const {name, date} = formState;

    const onSelectChange = (event: SelectChangeEvent) => {
        if (event.target.value === 'YES') return setIsEnable(true);
        setIsEnable(false);
    }

    const handleEdit = () => {
        setIsEditing(true);
    }

    const handleSave = () => {
        setIsEditing(false);
        dispatch(startUpdateQrCode({
            ...qrCodeRow,
            ...formState,
            enabled: isEnable
        }));
    }

    const handleDelete = () => {
        dispatch(startDeleteQrCodeWithDependencies(qrCodeRow.id));
    }

    const handleShow = () => {

    }

    const handleDownload = () => {

    }

    return (
        <TableRow sx={{'& > *': {borderBottom: 'unset'}}}>
            <TableCell>
                <IconButton
                    aria-label="expand row"
                    size="small"
                    onClick={() => handleOpenSubTable(!open)}
                >
                    {open ? <KeyboardArrowUpIcon/> : <KeyboardArrowDownIcon/>}
                </IconButton>
            </TableCell>
            <TableCell component="th" scope="row">
                <ConditionalTextField
                    name="name"
                    value={name}
                    onChange={onInputChange}
                    condition={isEditing}
                    styles={{width: '120px'}}
                />
            </TableCell>
            <TableCell align="center">
                {registriesLength}
            </TableCell>
            <TableCell align="center">
                <ConditionalTextField
                    name="date"
                    value={date}
                    onChange={onInputChange}
                    condition={isEditing}
                    styles={{width: '100px'}}
                />
            </TableCell>
            <TableCell align="center">
                {
                    isEditing
                        ? <Select variant={'standard'} value={isEnable ? 'YES' : 'NO'} onChange={onSelectChange}>
                            <MenuItem value="YES">Yes</MenuItem>
                            <MenuItem value="NO">No</MenuItem>
                        </Select>
                        : isEnable ? 'YES' : 'NO'
                }
            </TableCell>
            <TableCell align="center">
                {
                    isEditing
                        ? <IconButton onClick={handleSave}><Save/></IconButton>
                        : <QrCodeMenuOptions
                            handleEdit={handleEdit}
                            handleDelete={handleDelete}
                            handleShow={handleShow}
                            handleDownload={handleDownload}
                        />
                }

            </TableCell>
        </TableRow>
    )
}