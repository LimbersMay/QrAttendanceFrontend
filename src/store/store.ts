import {AnyAction, configureStore, ThunkAction} from '@reduxjs/toolkit';
import {authSlice} from "./auth";
import {groupSlice, qrCodeSlice, registrySlice} from "./qrAttendance";
import socketMiddleware from "./middlewares/socketMiddleware";
export const store = configureStore({
    reducer: {
        auth: authSlice.reducer,
        group: groupSlice.reducer,
        qrCode: qrCodeSlice.reducer,
        registry: registrySlice.reducer
    },
    middleware: [socketMiddleware]
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;

export type AppThunk<ReturnType = void> = ThunkAction<
    ReturnType,
    RootState,
    unknown,
    AnyAction
>
