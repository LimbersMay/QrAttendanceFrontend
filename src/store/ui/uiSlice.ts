import {createSlice} from '@reduxjs/toolkit';
import {RootState} from "../store";

interface uiInitialState {
    isQrCodeModalOpen: boolean;
    isRegistryModalOpen: boolean;
    isShowingQrCode: boolean;
    isTitleModalOpen: boolean;
}

const initialState: uiInitialState = {
    isQrCodeModalOpen: false,
    isRegistryModalOpen: false,
    isShowingQrCode: false,
    isTitleModalOpen: false
}

export const uiSlice = createSlice({
    name: 'ui',
    initialState,
    reducers: {
        onShowQrCode: (state) => {
            state.isShowingQrCode = true;
        },
        onHideQrCode: (state) => {
            state.isShowingQrCode = false;
        },
        onToggleQrCodeModal: (state) => {
            state.isQrCodeModalOpen = !state.isQrCodeModalOpen;
        },
        onToggleRegistryModal: (state) => {
            state.isRegistryModalOpen = !state.isRegistryModalOpen;
        },
        onToggleTitleModal: (state) => {
            state.isTitleModalOpen = !state.isTitleModalOpen;
        }
    }
});


// Action creators are generated for each case reducer function
export const { onToggleQrCodeModal, onShowQrCode, onHideQrCode, onToggleRegistryModal, onToggleTitleModal } = uiSlice.actions;
export const selectUi = (state: RootState) => state.ui;