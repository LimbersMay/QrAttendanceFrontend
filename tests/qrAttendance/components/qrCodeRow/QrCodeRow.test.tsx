import {render, screen} from "@testing-library/react";
import {QrCodeRow} from "../../../../src/qrAttendance/components/qrCodeRow/QrCodeRow";
import {qrCodes, withActiveGroupAndGroupsState, withQrCodesState} from "../../../fixtures/qrAttendanceStates";
import {registries} from "../../../fixtures/registryStates";
import {configureStore} from "@reduxjs/toolkit";
import {groupSlice, qrCodeSlice} from "../../../../src/store/qrAttendance";
import {Provider} from "react-redux";
import {uiSlice} from "../../../../src/store/ui/uiSlice";
import {initialState} from "../../../fixtures/uiStates";

const store = configureStore({
    reducer: {
        ui: uiSlice.reducer,
        qrCode: qrCodeSlice.reducer,
        group: groupSlice.reducer

    },
    preloadedState: {
        ui: {...initialState},
        qrCode: {...withQrCodesState},
        group: {...withActiveGroupAndGroupsState}
    }
})

describe('Tests for <QrCodeRow />', () => {
    test('Should match the snapshot', () => {

        const { container } = render(
            <Provider store={store}>
                <table>
                    <tbody>
                        <QrCodeRow qrCodeRow={qrCodes[0]} handleOpenSubTable={() => {}} open={false} registries={registries} />
                    </tbody>
                </table>
            </Provider>
        );

        expect(container).toMatchSnapshot();
    });

    test('should show the name of the qrCode', () => {
         render(
            <Provider store={store}>
                <QrCodeRow qrCodeRow={qrCodes[0]} handleOpenSubTable={() => {}} open={false} registries={registries} />
            </Provider>
        );

        expect(screen.getByText(qrCodes[0].name)).toBeTruthy();
    });
});
