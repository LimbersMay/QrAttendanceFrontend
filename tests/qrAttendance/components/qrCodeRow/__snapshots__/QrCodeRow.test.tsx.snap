// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <QrCodeRow /> Should match the snapshot 1`] = `
<div>
  <table>
    <tbody>
      <tr
        class="MuiTableRow-root css-2blr9s-MuiTableRow-root"
      >
        <td
          class="MuiTableCell-root MuiTableCell-sizeMedium css-bsph2u-MuiTableCell-root"
        >
          <button
            aria-label="expand row"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="KeyboardArrowDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </td>
        <th
          class="MuiTableCell-root MuiTableCell-sizeMedium css-bsph2u-MuiTableCell-root"
          scope="row"
        >
          Group A
        </th>
        <td
          class="MuiTableCell-root MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ce3qcx-MuiTableCell-root"
        >
          2
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1lu91p5-MuiTableCell-root"
        >
          01/01/2001
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ce3qcx-MuiTableCell-root"
        >
          NO
        </td>
        <td
          class="MuiTableCell-root MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ce3qcx-MuiTableCell-root"
        >
          <button
            aria-controls="long-menu"
            aria-haspopup="true"
            aria-label="more"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="MoreVertIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
